core/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -120,48 +123,65 @@ public SnapshotsService(Settings settings, ClusterService clusterService, IndexN
     }
 
     /**
+     * Retrieves list of snapshot ids that are present in a repository
+     *
+     * @param repositoryName repository name
+     * @return list of snapshot ids
+     */
+    public List<SnapshotId> snapshotIds(final String repositoryName) {
+        Repository repository = repositoriesService.repository(repositoryName);
+        assert repository != null; // should only be called once we've validated the repository exists
+        return repository.snapshots();
+    }
+
+    /**
      * Retrieves snapshot from repository
      *
-     * @param snapshotId snapshot id
+     * @param repositoryName  repository name
+     * @param snapshotId      snapshot id
      * @return snapshot
      * @throws SnapshotMissingException if snapshot is not found
      */
-    public SnapshotInfo snapshot(SnapshotId snapshotId) {
-        validate(snapshotId);
-        List<SnapshotsInProgress.Entry> entries = currentSnapshots(snapshotId.getRepository(), new String[]{snapshotId.getSnapshot()});
+    public SnapshotInfo snapshot(final String repositoryName, final SnapshotId snapshotId) {
+        List<SnapshotsInProgress.Entry> entries = currentSnapshots(repositoryName, Arrays.asList(snapshotId.getName()));
         if (!entries.isEmpty()) {
             return inProgressSnapshot(entries.iterator().next());
         }
-        return repositoriesService.repository(snapshotId.getRepository()).readSnapshot(snapshotId);
+        return repositoriesService.repository(repositoryName).readSnapshot(snapshotId);
     }
 
     /**
      * Returns a list of snapshots from repository sorted by snapshot creation date
      *
      * @param repositoryName repository name
+     * @param snapshotIds       snapshots for which to fetch snapshot information
+     * @param ignoreUnavailable if true, snapshots that could not be read will only be logged with a warning,
+     *                          if false, they will throw an error
      * @return list of snapshots
      */
-    public List<SnapshotInfo> snapshots(String repositoryName, boolean ignoreUnavailable) {
-        Set<SnapshotInfo> snapshotSet = new HashSet<>();
-        List<SnapshotsInProgress.Entry> entries = currentSnapshots(repositoryName, null);
+    public List<SnapshotInfo> snapshots(final String repositoryName, List<SnapshotId> snapshotIds, final boolean ignoreUnavailable) {
+        final Set<SnapshotInfo> snapshotSet = new HashSet<>();
+        final List<SnapshotId> snapshotIdsToIterate = new ArrayList<>(snapshotIds);
+        final List<SnapshotsInProgress.Entry> entries =
+            currentSnapshots(repositoryName, snapshotIds.stream().map(SnapshotId::getName).collect(Collectors.toList()));
         for (SnapshotsInProgress.Entry entry : entries) {
             snapshotSet.add(inProgressSnapshot(entry));
+            snapshotIdsToIterate.remove(entry.snapshot().getSnapshotId());
```

Comment:
here, we prioritize matching running snapshots to stored snapshots, but do it the other way around when we explicitly match in TransportGetSnapshotsAction?